### PR TITLE
Make optional Slack Hook service

### PR DESCRIPTION
**Pull request by nreal**
# Background
We need to make optional the slack hook service because this API should not fail when settings for Slack hook service is not deployed

# Additional changes
- Add DummySlackHooksService.cs when not exist Slack settings
- Slack hook service now has an `IHttpFactory `
- Improvements in Startup.cs

# Settings
If you want Slack integration, can add the key appsettings.json
```
"SlackHook": {
      "Url": "",
      "Text": ""
}

### CODE REVIEW

File: Doppler.Currency/SlackHookServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        public static IServiceCollection AddSlackHook(this IServiceCollection services)
+        {
+            services.AddSingleton<ISlackHooksService>(provider =>
+            {
+                var configuration = provider.GetService<IConfiguration>();
+                var section = configuration.GetSection("SlackHook");
+
+                if (!section.Exists()) 
+                    return new DummySlackHooksService(provider.GetService<ILogger<DummySlackHooksService>>());
```
Comment (andresmoschini) [inline Doppler.Currency/SlackHookServiceCollectionExtensions.cs:20]: Did you tried this?

```suggestion
                    return provider.GetService<DummySlackHooksService>();
```

File: Doppler.Currency/Program.cs [hunk 2]
```diff
@@ -15,6 +16,11 @@ public static void Main(string[] args)
             Host.CreateDefaultBuilder(args)
                 .UseSerilog((hostContext, loggerConfiguration) => 
                     loggerConfiguration.ReadFrom.Configuration(hostContext.Configuration))
+                .ConfigureAppConfiguration(configHost =>
+                {
+                    configHost.AddJsonFile("/run/secrets/appsettings.Secret.json", true);
+                    configHost.AddKeyPerFile("/run/secrets", true);
+                })
```
Comment (elsupergomez) [inline Doppler.Currency/Program.cs:23]: This is out of the scope of this PR. It is fine add it, but would better do it in another PR.

File: Doppler.Currency/Services/CurrencyHandler.cs [hunk 3]
```diff
@@ -39,8 +39,8 @@ public abstract class CurrencyHandler
             Exception e = null)
         {
             Logger.LogError(e ?? new Exception("Error getting HTML"),
-                $"Error getting HTML, title is not valid, please check HTML: {htmlPage}");
-            await SlackHooksService.SendNotification(HttpClient, $"Can't get currency from {currencyCode} currency code, please check Html in the log or if the date is holiday {dateTime.ToUniversalTime():yyyy-MM-dd}");
+                    $"Error getting HTML, title is not valid, please check HTML: {htmlPage}");
```
Comment (elsupergomez) [inline Doppler.Currency/Services/CurrencyHandler.cs:42]: Try use [structured logging](https://softwareengineering.stackexchange.com/questions/312197/benefits-of-structured-logging-vs-basic-logging) always.
Additional, again bad indentation.
```suggestion
                "Error getting HTML, title is not valid, please check HTML: {Page}", htmlPage);
```

File: Doppler.Currency/appsettings.json [hunk 4]
```diff
@@ -43,9 +43,10 @@
...
-  },
-  "SlackHook": {
-    "Url": "[SECRET KEY]",
-    "Text": "Doppler Currency Service can't get the currency today, Please check Html format"
   }
+  /* Add Property SlackHook if you want Slack integration
+    "SlackHook": {
+      "Url": "",
+      "Text": ""
+    }*/
```
Comment (elsupergomez) [inline Doppler.Currency/appsettings.json:51]: I feel that comment in the `appsettings.json` is not a good approach.
What do you think about leve el `Url` with a empty value as a not configured service?

File: Doppler.Currency/Startup.cs [hunk 5]
```diff
@@ -79,21 +75,6 @@ public void ConfigureServices(IServiceCollection services)
...
 
-        private void AddServiceSettings(IServiceCollection services)
-        {
-            var dofSettings = new CurrencySettings();
-            Configuration.GetSection("CurrencyCode:DofService").Bind(dofSettings);
-            services.AddSingleton(dofSettings);
-
-            var bnaSettings = new CurrencySettings();
-            Configuration.GetSection("CurrencyCode:BnaService").Bind(bnaSettings);
-            services.AddSingleton(bnaSettings);
```
Comment (elsupergomez) [inline Doppler.Currency/Startup.cs:90]: How are configured this services now?
I do not fine any other reference in this PR.

File: Doppler.Currency/SlackHookServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    public static class SlackHookServiceCollectionExtensions
+    {
+        public static IServiceCollection AddSlackHook(this IServiceCollection services)
+        {
+            services.AddSingleton<ISlackHooksService>(provider =>
+            {
+                var configuration = provider.GetService<IConfiguration>();
+                var section = configuration.GetSection("SlackHook");
+
+                if (!section.Exists())
```
Comment (elsupergomez) [inline Doppler.Currency/SlackHookServiceCollectionExtensions.cs:19]: What do you think about check if the `Uri` value is null or empty instead of check if the section exist?
And leve the `appsettings.json` with empty settings.
Maybe use a Service Factory class could be a good alternative.

**Comment by nreal** [comment]
Close in favor https://github.com/FromDoppler/doppler-currency/pull/17

File: Doppler.Currency/appsettings.json (hunk 4 above)
Comment (andresmoschini) [reply Doppler.Currency/appsettings.json:51]: I think that it is _more correct_  avoiding the property _SlackHook_ when we do not want to use it. It is more explicit than assuming that behavior when the URL is empty.

But I am not entirely happy, this comment is a smell that our decision is not entirely well. 

@elsupergomez what do you think a different approach, something like this:

```json
    "SlackHook": {
      "Enabled": false,
      "Url": "",
      "Text": ""
    }
```
